test/langtools/tools/javac/sealed/ValidateJarWithSealedAndRecord.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8282446 8282508
+ * @summary Jar validation fails when sealed classes and records are involved
+ * @modules jdk.compiler
+ *          jdk.jartool/sun.tools.jar
+ * @run main ValidateJarWithSealedAndRecord
+ */
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.util.ArrayList;
+import java.util.List;
+import javax.tools.DiagnosticCollector;
+import javax.tools.JavaFileObject;
+import javax.tools.StandardJavaFileManager;
+import javax.tools.StandardLocation;
+import javax.tools.ToolProvider;
+
+public class ValidateJarWithSealedAndRecord {
+
+    public static void main(String args[]) throws Exception {
+        ValidateJarWithSealedAndRecord theTest = new ValidateJarWithSealedAndRecord();
+        theTest.run();
+    }
+
+    void run() throws Exception {
+        generateFilesNeeded();
+    }
+
+    void writeFile(String f, String contents) throws IOException {
+        PrintStream s = new PrintStream(new FileOutputStream(f));
+        s.println(contents);
+        s.close();
+    }
+
+    void generateFilesNeeded() throws Exception {
+        sun.tools.jar.Main jarGenerator = new sun.tools.jar.Main(System.out, System.err, "jar");
```

Comment:
You can call `java.util.spi.ToolProvider.findFirst("jar")` instead of the internal `sun.tools.jar.Main` class.  This way the test does not need access to the module-private package.